linera-chain/src/data_types.rs

```diff
@@ -587,16 +587,62 @@ impl<'a> SignatureAggregator<'a> {
     }
 }
 
+fn is_ordered(values: &[(ValidatorName, Signature)]) -> bool {
+    values.windows(2).all(|pair| pair[0].0 <= pair[1].0)
+}
+
+impl<'de> Deserialize<'de> for Certificate {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct CertificateVisitor;
+
+        impl<'de> serde::de::Visitor<'de> for CertificateVisitor {
+            type Value = Certificate;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("struct Certificate")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
```

Comment:
Good job navigating the visitor pattern! This being said, to keep things simple, we usually delegate the deserialization to a temporary data structure instead. (Look for `#[serde(rename = "PublicKey")]` in `crypto.rs` for an example.)